### PR TITLE
[bug]:corrected the search functionality

### LINKED ISSUE #9738: [BUG] Search page rolls back to initial state whenever a new search filter is added/removed.
### Has this bug been raised before?

- [X] I have checked "open" AND "closed" issues and this is not a duplicate

### Where did you find this bug?

Production

### Version of BioDrop (for example "v1.2.3")

v2.84.4

### Description

On the search page, when multiple filters are selected whenever a new filter is added/removed the display data rolls back to the initial state for a moment and then displays the new data.
For example if I have selected react, then the result will be shown for react then if i also select python then for a few seconds i will see the initial unfiltered search result and then the result for react+python filter

### Screenshots

_No response_

### Do you want to work on this issue?

Yes

### If "yes" to above, please explain how you would technically implement this

_No response_

### PR DESCRIPTION
**Pull request by jatingodnani**
closes #9738 
- [*]  My code follows the code style of this project.

- [* ] My change requires changes to the documentation.
- [ *] I have updated the documentation accordingly.
- [* ] All new and existing tests passed.
- [ *] This PR does not contain plagiarized content.
- [ *] The title of my pull request is a short description of the requested changes.

Added loader indicator so that if any search, intially  at the query time it shows the recently updated user but i have done some changes or added to the code so it showing loading indicatior, not the recently updated users.

### CODE REVIEW

File: pages/search.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ import {
 import { PROJECT_NAME } from "@constants/index";
 
 async function fetchUsersByKeyword(keyword) {
+  console.log(keyword)
```
Comment (sital002) [inline pages/search.js:21]:
```suggestion
```

File: pages/search.js [hunk 2]
```diff
@@ -195,6 +204,7 @@ export default function Search({
   const indexOfFirstUser = indexOfLastUser - usersPerPage;
   const visibleUsers = users.slice(indexOfFirstUser, indexOfLastUser);
 
+
```
Comment (eddiejaoude) [inline pages/search.js:207]: Extra blank line not needed here, please check your Prettier settings

**Review by eddiejaoude — CHANGES_REQUESTED** [review]
Thank you for the contribution 👍 

I have not tested it yet, but please check your Prettier settings, so it is easier to review the "actual" lines changed

**Comment by jatingodnani** [comment]
@eddiejaoude review my code

File: pages/search.js [hunk 3]
```diff
@@ -150,10 +134,34 @@ export default function Search({
       );
       fetchUsers(inputValue);
     }, 500);
-
-    return () => clearTimeout(timer);
+    console.log(inputValue);
```
Comment (badrivlog) [inline pages/search.js:137]: please remove this debugging code

File: pages/search.js [hunk 4]
```diff
@@ -150,10 +134,34 @@ export default function Search({
...
+      setNotFound();
+      setUsers(data.users.sort(() => Math.random() - 0.5));
+      setCurrentPage(1);
+    } catch (err) {
+      setNotFound(err.message);
+      setUsers([]);
+    } finally {
+      setLoading(false);
+    }
+  }
```
Comment (badrivlog) [inline pages/search.js:164]: I'm not sure why you moved this function outside from useEffect?

File: pages/search.js [hunk 5]
```diff
@@ -239,28 +247,39 @@ export default function Search({
...
+          <h2 className="mt-10 mb-4 text-2xl font-bold">
+            Recently updated profiles
+          </h2>
+        )}
 
         {notFound && <Alert type="error" message={notFound} />}
         <ul
           role="list"
-          className="grid grid-cols-1 gap-6 sm:grid-cols-2 lg:grid-cols-3"
+          className="grid grid-cols-1 gap-6 sm:grid-cols-2 lg:grid-cols-3 "
```
Comment (badrivlog) [inline pages/search.js:259]:
```suggestion
          className="grid grid-cols-1 gap-6 sm:grid-cols-2 lg:grid-cols-3"
```

File: pages/search.js [hunk 6]
```diff
@@ -239,28 +247,39 @@ export default function Search({
...
-              </li>
-            ))}
-
-          {users.length > usersPerPage &&
-            visibleUsers.map((user) => (
-              <li key={user.username}>
-                <UserHorizontal profile={user} input={inputValue} />
-              </li>
-            ))}
+          {visibleUsers.length < usersPerPage && inputValue == ""
```
Comment (badrivlog) [inline pages/search.js:261]:
```suggestion
          {visibleUsers.length < usersPerPage && inputValue === ""
```

File: pages/search.js (hunk 4 above)
Comment (jatingodnani) [reply pages/search.js:164]: Beacuse we are using fetchuser only in timer,to make code cleaner,it doesnt effect any function

**Comment by eddiejaoude** [comment]
Seems like the search tests are failing

**Comment by jatingodnani** [comment]
Why this is failing,in my local host it is working fine
Is there any reason

**Comment by eddiejaoude** [comment]
> Why this is failing,in my local host it is working fine Is there any reason

the automated tests pass for you locally?

**Comment by jatingodnani** [comment]
I have change some ui,which is working fine in my local
Can you pls test my code,then let me know why this is failing

**Comment by eddiejaoude** [comment]
> Why this is failing,in my local host it is working fine

can you please share the results of running the tests locally

> I have change some ui,which is working fine in my local Can you pls test my code,then let me know why this is failing

you can see the failing tests in the GitHub Action

**Comment by jatingodnani** [comment]
@eddiejaoude help me to find why 5th case is failing

**Comment by badrivlog** [comment]
@jatingodnani please resolve the merge conflicts

**Comment by eddiejaoude** [comment]
It seems like the test checks are expected different results before as to now

```js
1) [chromium] › search.spec.js:22:5 › Search works correctly ─────────────────────────────────────

    Error: Timed out 5000ms waiting for expect(locator).toHaveCount(expected)

    Locator: locator('main li')
    Expected: 1
    Received: 0
```

Please run the tests and check your changes.

https://github.com/EddieHubCommunity/BioDrop/actions/runs/6913193428/job/18823909700?pr=9788

**Comment by adityaraute** [comment]
@jatingodnani  were you able to resolve the conflicts?

**Comment by jatingodnani** [comment]
Yes,i solved this problem ,but i am not understand why my test case is failing

File: pages/search.js [hunk 7]
```diff
@@ -150,10 +134,34 @@ export default function Search({
       );
       fetchUsers(inputValue);
     }, 500);
-
-    return () => clearTimeout(timer);
+    
+    return () => {
+      clearTimeout(timer);
+    };
```
Comment (eddiejaoude) [inline pages/search.js:140]: Does the coding style need to be changed here?

File: pages/search.js [hunk 8]
```diff
@@ -239,28 +247,39 @@ export default function Search({
...
+                  <UserHorizontal profile={user} input={inputValue} />
+                </li>
+              ))}
         </ul>
 
+        {loading && (
+          <div className="flex items-center justify-center mt-8 w-[97%]">
+            <div
+              class=" animate-spin rounded-full border-opacity-25 h-10 w-10 border-4
+                 border-t-blue-500 "
```
Comment (eddiejaoude) [inline pages/search.js:279]:
```suggestion
                 border-t-blue-500"
```

File: pages/search.js [hunk 9]
```diff
@@ -239,28 +247,39 @@ export default function Search({
...
+            : users.length > usersPerPage &&
+              visibleUsers.map((user) => (
+                <li key={user.username}>
+                  <UserHorizontal profile={user} input={inputValue} />
+                </li>
+              ))}
         </ul>
 
+        {loading && (
+          <div className="flex items-center justify-center mt-8 w-[97%]">
```
Comment (eddiejaoude) [inline pages/search.js:276]: `97%` seems a bit strange?

**Comment by eddiejaoude** [comment]
1. please resolve conflicts
2. test is failing because the expect results has changed

From the GitHub Action, you can see...

```
Expected: 9
Received: 27
```

**Comment by SaraJaoude** [comment]
Given the length of time that this PR has been open and maintainer comments have not been addressed, I will close this to give someone else in the community an opportunity to work on it.

**Pull request closed without merging**